lib/client-api/src/decorators.test.ts

```diff
@@ -42,6 +42,50 @@ describe('client-api.decorators', () => {
     expect(contexts.map((c) => c.k)).toEqual([0, 3, 2, 1]);
   });
 
+  it('does not recreate decorated story functions each time', () => {
+    const decoratedStories = [];
+    const decorators = [
+      (s, c) => {
+        decoratedStories.push = s;
+        return s();
+      },
+    ];
+    const decorated = defaultDecorateStory(() => 0, decorators);
+
+    decorated(makeContext());
+    decorated(makeContext());
+    expect(decoratedStories[0]).toBe(decoratedStories[1]);
+  });
+
+  // NOTE: important point--this test would not work if we called `decoratedOne` twice simultaneously
+  // both story functions would receive {story: 2}. The assumption here is that we'll never render
+  // the same story twice at the same time.
+  it('does not interleave contexts if two decorated stories are call simultaneously', async () => {
+    const contexts = [];
+    let resolve;
+    const fence = new Promise((r) => {
+      resolve = r;
+    });
+    const decorators = [
+      async (s, c) => {
+        // The fence here simulates async-ness in react rendering an element (`<S />` doesn't run `S()` straight away)
+        await fence;
+        s();
+      },
+    ];
+    const decoratedOne = defaultDecorateStory((c) => contexts.push(c), decorators);
+    const decoratedTwo = defaultDecorateStory((c) => contexts.push(c), decorators);
+
+    decoratedOne(makeContext({ story: 1 }));
+    decoratedTwo(makeContext({ story: 2 }));
+
+    resolve();
+    await fence;
+
+    expect(contexts[0].story).toBe(1);
+    expect(contexts[1].story).toBe(2);
```

Comment:
I'd really like to get rid of the `.story` property at some point. Can we instead test for `.id`? AFAIK the `.story` property we don't actually use, and is just there to not break backwards compatibility.

Reply:
Oh, this `.story` property is just one I made for this test. Let's call it something else.